packages/signers/src/sign-offchain-message.ts

```diff
@@ -0,0 +1,165 @@
+ *
+ * This function delegates to the {@link partiallySignOffchainMessageWithSigners} function
+ * in order to extract signers from the offchain message and sign it.
+ *
+ * @example
+ * ```ts
+ * const mySignedOffchainMessageEnvelope = await signOffchainMessageWithSigners(myOffchainMessage);
+ *
+ * // With additional config.
+ * const mySignedOffchainMessageEnvelope = await signOffchainMessageWithSigners(myOffchainMessage, {
+ *     abortSignal: myAbortController.signal,
+ * });
+ *
+ * // We now know the offchain message is fully signed.
+ * mySignedOffchainMessageEnvelope satisfies FullySignedOffchainMessageEnvelope;
+ * ```
+ *
+ * @see {@link partiallySignOffchainMessageWithSigners}
+ */
+export async function signOffchainMessageWithSigners(
+    offchainMessage: OffchainMessage,
+    config?: MessagePartialSignerConfig,
+): Promise<FullySignedOffchainMessageEnvelope & OffchainMessageEnvelope> {
+    const signedOffchainMessageEnvelope = await partiallySignOffchainMessageWithSigners(offchainMessage, config);
+    assertIsFullySignedOffchainMessageEnvelope(signedOffchainMessageEnvelope);
+    return signedOffchainMessageEnvelope;
+}
+
+/**
+ * Identifies each provided {@link MessageSigner} and categorizes them into their respective types.
+ * When a signer implements multiple interfaces, it will try to used to most powerful interface but
+ * fall back to the least powerful interface when necessary.
+ *
+ * For instance, if a signer implements {@link MessageSigner} and {@link MessageModifyingSigner},
+ * it will be categorized as a `MessageModifyingSigner`.
+ */
+function categorizeMessageSigners(signers: readonly MessageSigner[]): Readonly<{
+    modifyingSigners: readonly MessageModifyingSigner[];
+    partialSigners: readonly MessagePartialSigner[];
+}> {
+    // Identify the modifying signers from the other signers.
+    const modifyingSigners = identifyMessageModifyingSigners(signers);
+
+    // Use any remaining signers as partial signers.
+    const partialSigners = signers
+        .filter(isMessagePartialSigner)
+        .filter(signer => !(modifyingSigners as typeof signers).includes(signer));
+
+    return Object.freeze({ modifyingSigners, partialSigners });
+}
+
+/** Identifies the best signers to use as MessageModifyingSigners, if any */
+function identifyMessageModifyingSigners(
+    signers: readonly (MessageModifyingSigner | MessagePartialSigner)[],
+): readonly MessageModifyingSigner[] {
+    // Ensure there are any MessageModifyingSigner in the first place.
+    const modifyingSigners = signers.filter(isMessageModifyingSigner);
+    if (modifyingSigners.length === 0) return [];
+
+    // Prefer modifying signers that do not offer partial signing.
+    const nonPartialSigners = modifyingSigners.filter(signer => !isMessagePartialSigner(signer));
+    if (nonPartialSigners.length > 0) return nonPartialSigners;
+
+    // Otherwise, choose only one modifying signer (whichever).
+    return [modifyingSigners[0]];
```

Comment:
Perfect!

Reply:
I can't think of any use-cases for modifying these messages when signing but there is likely some that I don't see and I think it would be nice to support it for consistency with the rest of the signer API.

Do you have any concerns with this approach?

Reply:
I guess one example is if your v0 message can be displayed with Restricted ASCII but you gave it the message format ASCII, a software wallet that knows it's using a Ledger account might swap the message format before sending it to Ledger.

Also the SIWS spec ([or at least Phantom's docs for it](https://github.com/phantom/sign-in-with-solana?tab=readme-ov-file#specification)) have some fields the wallet is expected to determine if they're not provided, so if you were using this for SIWS then you'd expect the wallet to modify the nested message in those cases.